hs/src/Reach/Eval/Core.hs

```diff
@@ -2270,88 +2353,93 @@ instDefaultArgs env err formals = \case
-               , sco_must_ret = RS_MayBeEmpty
-               , sco_while_vars = Nothing
-               , sco_penvs = clo_penvs
-               , sco_cenv = clo_cenv
-               , sco_use_strict = clo_use_strict
-               })
-      m <- readSt st_mode
-      arg_env <-
-        locStMode (pure_mode m) $
-          locSco clo_sco $
-            instDefaultArgs mempty err formals randvs
-      at <- withAt id
-      clo_sco' <- locSco clo_sco $ sco_update arg_env
-      (body_lifts, (SLStmtRes clo_sco'' rs)) <-
-        captureLifts $
-          withFrame (SLC_CloApp at clo_at mname) $
-            locAt body_at $
-              locSco clo_sco' $
-                evalStmt body
-      let no_prompt (lvl, v) = do
-            let lifts' =
-                  case body_lifts of
-                    body_lifts' Seq.:|> (DLS_Return _ the_ret_label _the_val)
-                      | the_ret_label == ret ->
-                        --- We don't check that the_val is v, because
-                        --- we're relying on the invariant that there
-                        --- was only one Return... this should be
-                        --- true, but if something changes in the
-                        --- future, this might be a place that an
-                        --- error could be introduced.
-                        body_lifts'
-                    _ ->
-                      return $ DLS_Prompt body_at (Left ret) body_lifts
-            saveLifts lifts'
-            return $ SLAppRes clo_sco'' $ (lvl, v)
-      case rs of
-        [] -> no_prompt $ public $ SLV_Null body_at "clo app"
-        [(_, _, x, False)] -> no_prompt $ x
-        (_, _, (xlvl, xv), _) : more -> do
-          let msvs = map (\(_a, _b, c, _d) -> c) more
-          let mlvls = map fst msvs
-          let mvs = map snd msvs
-          let lvl = mconcat $ xlvl : mlvls
-          -- Note: This test might be too expensive, so try it
-          let all_same = False && all (== xv) mvs
-          case all_same of
-            True -> no_prompt $ (lvl, xv)
-            False -> do
-              let go (r_at, rmi, (_, rv), _) = do
-                    (rlifts, (rty, rda)) <-
-                      captureLifts $ locAt r_at $ compileTypeOf rv
-                    let retsm =
-                          case rmi of
-                            Nothing -> mempty
-                            Just ri -> M.singleton ri (rlifts, rda)
-                    return $ (retsm, rty)
-              (retsms, tys) <- unzip <$> mapM go rs
-              let retsm = mconcat retsms
-              r_ty <- locAt body_at $ typeMeets_d tys
-              let dv = DLVar body_at Nothing r_ty ret
-              saveLift $ DLS_Prompt body_at (Right (dv, retsm)) body_lifts
-              return $ SLAppRes clo_sco'' (lvl, (SLV_DLVar dv))
+    SLV_Clo clo_at sc -> evalApplyClosureVals clo_at sc randvs
+    SLV_CloTyped clo_at sc _ -> evalApplyClosureVals clo_at sc randvs
```

Comment:
I think that this should apply the refinements

Reply:
If they are present and potentially check the types, although not call with `DLVar`s